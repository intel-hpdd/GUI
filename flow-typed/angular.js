// @flow

declare module angular {
  declare class $scope {
    $new():$scope;
    $digest():void;
  }
  declare type $scopeT = $scope;
  declare class Module {
    provider(name:string, providerType:Function):Module;
    factory(name:string, providerFunction:Function):Module;
    controller(name:string, constructor:Function):Module;
    service(name:string, constructor:Function):Module;
    config(configFn:Function):Module;
    run(initializationFn:Function):Module;
    directive(name:string, directiveFactory:Function):Module;
    component(name:string, options:Object):Module;
    filter(name:string, filterFactory:Function):Module;
    value(name:string, object:any):Module;
    constant(name:string, object:any):Module;
    name:string;
  }
  declare type $compileT = (el:(string | HTMLElement)) => (scope:$scope) => HTMLElement[];
  declare module.exports: {
    module(name:string, dependencies: string[]):Module;
    bootstrap(element:(HTMLElement | Document), modules:string[], config:Object):void;
    merge(...rest:Object[]):Object;
    element(el:(string | HTMLElement)): {
      clone: () => HTMLElement
    };
  }
}
