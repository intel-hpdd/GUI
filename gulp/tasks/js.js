//
// INTEL CONFIDENTIAL
//
// Copyright 2013-2016 Intel Corporation All Rights Reserved.
//
// The source code contained or described herein and all documents related
// to the source code ("Material") are owned by Intel Corporation or its
// suppliers or licensors. Title to the Material remains with Intel Corporation
// or its suppliers and licensors. The Material contains trade secrets and
// proprietary and confidential information of Intel or its suppliers and
// licensors. The Material is protected by worldwide copyright and trade secret
// laws and treaty provisions. No part of the Material may be used, copied,
// reproduced, modified, published, uploaded, posted, transmitted, distributed,
// or disclosed in any way without Intel's prior express written permission.
//
// No license under any patent, copyright, trade secret or other intellectual
// property right is granted to or conferred upon you by disclosure or delivery
// of the Materials, either expressly, by implication, inducement, estoppel or
// otherwise. Any license under such intellectual property rights must be
// express and approved by Intel in writing.

'use strict';

var paths = require('../paths.json');
var gulp = require('gulp');
var babel = require('gulp-babel');
var sourcemaps = require('gulp-sourcemaps');
var destDir = require('../dest-dir');
var relativeSourcemapsSource = require('gulp-relative-sourcemaps-source');

function jsDeps (fn) {
  return gulp.src(paths.js.deps, {
    since: gulp.lastRun(fn),
    base: '.'
  });
}

exports.jsDepsDev = function jsDepsDev () {
  return jsDeps(jsDepsDev)
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.symlink('static/chroma_ui', { cwd: destDir }));
};

exports.jsDepsProd = function jsDepsProd () {
  return jsDeps(jsDepsProd)
  .pipe(gulp.dest('./dest'));
};

function socketWorker (fn) {
  return gulp.src(paths.js.socketWorker, {
    since: gulp.lastRun(fn)
  })
  .pipe(sourcemaps.init({ loadMaps: true }))
  .pipe(sourcemaps.write('.'));
}

exports.socketWorkerDev = function socketWorkerDev () {
  return socketWorker(socketWorkerDev)
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.symlink('static/chroma_ui', { cwd: destDir }));
};

exports.socketWorkerProd = function socketWorkerProd () {
  return socketWorker(socketWorkerProd)
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.dest('./dist'));
};

function systemConfig (fn) {
  return gulp.src(paths.js.config, {
    since: gulp.lastRun(fn),
    base: '.'
  });
}

exports.systemConfigDev = function systemConfigDev () {
  return systemConfig(systemConfigDev)
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.symlink('static/chroma_ui', { cwd: destDir }));
};

var babelDev = babel.bind(null, {
  presets: [],
  babelrc: false,
  plugins: [
    'check-es2015-constants',
    'transform-flow-strip-types',
    'transform-class-properties',
    'transform-object-rest-spread',
    'transform-async-to-generator',
    'transform-es2015-classes',
    'transform-es2015-modules-systemjs'
  ]
});

var babelProd = babel.bind(null, {
  presets: ['es2015'],
  babelrc: false,
  plugins: [
    'transform-flow-strip-types',
    'transform-class-properties',
    'transform-object-rest-spread',
    [
      'angularjs-annotate',
      {
        explicitOnly: true
      }
    ],
    'transform-es2015-modules-systemjs'
  ]
});

function jsSource (fn) {
  return gulp.src(paths.js.source, {
    since: gulp.lastRun(fn),
    base: '.'
  })
  .pipe(sourcemaps.init());
}

exports.jsSourceDev = function jsSourceDev () {
  return jsSource(jsSourceDev)
  .pipe(babelDev())
  .pipe(sourcemaps.write({ sourceRoot: '/static/chroma_ui' }))
  .pipe(gulp.dest('./dest'))
  .pipe(gulp.symlink('static/chroma_ui', { cwd: destDir }));
};

exports.jsSourceProd = function jsSourceProd () {
  return jsSource(jsSourceProd)
  .pipe(babelProd())
  .pipe(relativeSourcemapsSource({dest: 'dest'}))
  .pipe(sourcemaps.write({ sourceRoot: '' }))
  .pipe(gulp.dest('./dest'));
};

exports.jsTest = function jsTest () {
  return gulp.src(paths.js.tests, {
    since: gulp.lastRun(jsTest),
    base: '.'
  })
  .pipe(sourcemaps.init())
  .pipe(babelDev())
  .pipe(sourcemaps.write({ sourceRoot: '' }))
  .pipe(gulp.dest('./dest'));
};

exports.jsTestFixtures = function jsTestFixtures () {
  return gulp.src(paths.js.testFixtures, {
    since: gulp.lastRun(jsTestFixtures),
    base: '.'
  })
  .pipe(gulp.dest('./dest'));
};

exports.jsTestDeps = function jsTestDeps () {
  return gulp.src(paths.js.testDeps, {
    since: gulp.lastRun(jsTestDeps),
    base: '.'
  })
  .pipe(gulp.dest('./dest'));
};
