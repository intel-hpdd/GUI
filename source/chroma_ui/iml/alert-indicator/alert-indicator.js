//
// INTEL CONFIDENTIAL
//
// Copyright 2013-2015 Intel Corporation All Rights Reserved.
//
// The source code contained or described herein and all documents related
// to the source code ("Material") are owned by Intel Corporation or its
// suppliers or licensors. Title to the Material remains with Intel Corporation
// or its suppliers and licensors. The Material contains trade secrets and
// proprietary and confidential information of Intel or its suppliers and
// licensors. The Material is protected by worldwide copyright and trade secret
// laws and treaty provisions. No part of the Material may be used, copied,
// reproduced, modified, published, uploaded, posted, transmitted, distributed,
// or disclosed in any way without Intel's prior express written permission.
//
// No license under any patent, copyright, trade secret or other intellectual
// property right is granted to or conferred upon you by disclosure or delivery
// of the Materials, either expressly, by implication, inducement, estoppel or
// otherwise. Any license under such intellectual property rights must be
// express and approved by Intel in writing.

  angular.module('alertIndicator')
    .factory('alertMonitor',
      function alertMonitorFactory (socketStream) {
        return function alertMonitor () {
          var stream = socketStream('/alert/', {
            jsonMask: 'objects(affected,message)',
            qs: {
              limit: 0,
              active: true
            }
          });

          var s2 = stream
            .pluck('objects');

          s2.destroy = stream.destroy.bind(stream);

          return s2;
        };
      })
    .directive('recordState', function recordState (STATE_SIZE, $exceptionHandler, localApply) {
      return {
        scope: {
          recordId: '=',
          displayType: '=',
          alertStream: '='
        },
        restrict: 'E',
        replace: true,
        templateUrl: 'iml/alert-indicator/assets/html/alert-indicator.html',
        link: function link (scope) {
          var isOpen = false;

          var recordState = scope.recordState = {
            alerts: [],
            messageDifference: [],
            onToggle: function onToggle (state) {
              if (state === 'closed') {
                scope.recordState.clearMessageRecord();
                isOpen = false;
              } else {
                isOpen = true;
              }
            },
            isOpen: function isPopoverOpen () {
              return isOpen;
            },
            /**
             * Indicates if the command contains alerts (the directive is in an alert state).
             * @returns {Boolean}
             */
            isInErrorState: function isInErrorState () {
              return recordState.alerts.length > 0;
            },
            /**
             * Clears the message record and difference arrays.
             */
            clearMessageRecord: function clearMessageRecord () {
              scope.recordState.messageDifference = [];
              scope.recordState.messageRecord = [];
            },
            /**
             * Retrieves the tool tip message based on the number of alerts.
             * @returns {String}
             */
            getTooltipMessage: function getTooltipMessage () {
              var messageMap = {
                '0': 'No alerts.',
                '1': '1 alert message. Click to review details.',
                'other': '{} alert messages. Click to review details.'
              };

              return _.pluralize(recordState.alerts.length, messageMap);
            },
            showLabel: function showLabel () {
              return scope.displayType === STATE_SIZE.MEDIUM;
            }
          };

          var propertyStream = scope.alertStream.property();

          var indexOfRecord = fp.invokeMethod('indexOf', [ scope.recordId ]);
          var recordFound = fp.flow(fp.eqFn(fp.identity, indexOfRecord, -1), fp.not);

          propertyStream
            .map(fp.filter(fp.flow(fp.lensProp('affected'), recordFound)))
            .map(fp.map(fp.lensProp('message')))
            .stopOnError(fp.curry(1, $exceptionHandler))
            .tap(function setAlerts (alerts) {
              recordState.messageDifference = fp.difference(recordState.alerts, alerts);
            })
            .tap(fp.lensProp('alerts').set(fp.__, recordState))
            .each(localApply.bind(null, scope));

          scope.$on('$destroy', function onDestroy () {
            propertyStream.destroy();

            recordState = propertyStream = null;
          });
        }
      };
    });


