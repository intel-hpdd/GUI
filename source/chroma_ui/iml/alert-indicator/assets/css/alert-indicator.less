.record-state {
  .icon-wrap {
    position: relative;
  }

  margin-right: 10px;

  i {
    cursor: pointer;

    &.fa-exclamation-circle {
      color: @brand-danger;
    }

    &.fa-check-circle {
      color: @brand-success;
    }
  }

  .popover {
    width: 260px;

    > h3 {
      text-align: center;
    }

    accordion ul {
      padding-left: 10px;
    }
  }

  .state-label {
    margin-left: 10px;
  }
}
