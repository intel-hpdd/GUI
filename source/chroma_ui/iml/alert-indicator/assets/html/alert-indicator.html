<span class="record-state">
  <span class="icon-wrap">
    <i class="fa activate-popover"
       ng-class="{'fa-exclamation-circle': ctrl.hasAlerts(), 'fa-check-circle': !ctrl.hasAlerts() }"
       tooltip-template="'iml/alert-indicator/assets/html/state-label.html'"
       tooltip-placement="right"></i>
    <iml-popover placement="bottom" title="Alerts"
                 on-toggle="ctrl.onToggle(state)" ng-if="ctrl.hasAlerts()">
      <ul>
        <li ng-repeat="alert in ctrl.alerts">{{alert}}</li>
      </ul>
    </iml-popover>
  </span>
  <ng-pluralize class="state-label" ng-if="ctrl.showLabel()" count="ctrl.alerts.length"
                when="{
    0: 'No Issues',
    1: '{{ ctrl.alerts[0] }}',
    'other': '{} Issues'}">
  </ng-pluralize>
</span>
