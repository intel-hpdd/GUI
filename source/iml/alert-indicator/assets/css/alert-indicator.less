.record-state {
  .icon-wrap {
    position: relative;
  }

  width: 12px;
  height: 12px;
  margin-right: 10px;

  i {
    cursor: pointer;

    &.fa-exclamation-circle {
      color: @brand-danger;
    }

    &.fa-check-circle {
      color: @brand-success;
    }
  }

  .popover {
    max-width: 260px;

    > h3 {
      text-align: center;
    }

    accordion ul {
      padding-left: 10px;
    }
  }

  .state-label {
    margin-left: 10px;
  }
}
