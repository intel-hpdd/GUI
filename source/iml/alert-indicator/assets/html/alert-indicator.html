<span class="record-state">
  <span class="icon-wrap tooltip-container tooltip-hover">
    <i class="fa activate-popover"
       ng-class="{'fa-exclamation-circle': ctrl.hasAlerts(), 'fa-check-circle': !ctrl.hasAlerts() }"
    >
      <iml-tooltip size="ctrl.alerts.length === 0 ? 'small' : 'medium'" direction="right">
        <ng-pluralize count="ctrl.alerts.length"
                      when="{
          0: 'No Issues',
          1: '{{ ctrl.alerts[0] }}',
          'other': '{} Issues'}">
        </ng-pluralize>
      </iml-tooltip>
    </i>
    <iml-popover placement="bottom" title="Alerts"
                 on-toggle="ctrl.onToggle(state)" ng-if="ctrl.hasAlerts()">
      <ul>
        <li ng-repeat="alert in ctrl.alerts">{{alert}}</li>
      </ul>
    </iml-popover>
  </span>
  <ng-pluralize class="state-label" ng-if="ctrl.showLabel()" count="ctrl.alerts.length"
                when="{
    0: 'No Issues',
    1: '{{ ctrl.alerts[0] }}',
    'other': '{} Issues'}">
  </ng-pluralize>
</span>
