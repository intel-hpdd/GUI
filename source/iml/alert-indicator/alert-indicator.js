//
// INTEL CONFIDENTIAL
//
// Copyright 2013-2016 Intel Corporation All Rights Reserved.
//
// The source code contained or described herein and all documents related
// to the source code ("Material") are owned by Intel Corporation or its
// suppliers or licensors. Title to the Material remains with Intel Corporation
// or its suppliers and licensors. The Material contains trade secrets and
// proprietary and confidential information of Intel or its suppliers and
// licensors. The Material is protected by worldwide copyright and trade secret
// laws and treaty provisions. No part of the Material may be used, copied,
// reproduced, modified, published, uploaded, posted, transmitted, distributed,
// or disclosed in any way without Intel's prior express written permission.
//
// No license under any patent, copyright, trade secret or other intellectual
// property right is granted to or conferred upon you by disclosure or delivery
// of the Materials, either expressly, by implication, inducement, estoppel or
// otherwise. Any license under such intellectual property rights must be
// express and approved by Intel in writing.

import angular from 'angular';
import * as fp from 'intel-fp';

const viewLens = fp.flow(fp.lensProp, fp.view);
import alertIndicatorTemplate from './assets/html/alert-indicator.html!text';

export function RecordStateCtrl ($scope, STATE_SIZE, propagateChange) {
  'ngInject';

  const ctrl = angular.extend(this, {
    alerts: [],
    hasAlerts () {
      return ctrl.alerts.length > 0;
    },
    showLabel: function showLabel () {
      return ctrl.displayType === STATE_SIZE.MEDIUM;
    }
  });

  const viewer$ = ctrl.alertStream();
  const indexOfRecord = fp.invokeMethod('indexOf', [ctrl.recordId]);
  const recordFound = fp.flow(fp.eqFn(fp.identity, indexOfRecord, -1), fp.not);

  const p = propagateChange($scope, ctrl, 'alerts');

  viewer$
    .map(fp.filter(fp.flow(viewLens('affected'), recordFound)))
    .map(fp.map(viewLens('message')))
    .through(p);

  $scope.$on('$destroy', viewer$.destroy.bind(viewer$));
}

export const recordStateDirective = () => {
  'ngInject';

  return {
    scope: {},
    bindToController: {
      recordId: '=',
      displayType: '=',
      alertStream: '='
    },
    controller: 'RecordStateCtrl',
    controllerAs: 'ctrl',
    restrict: 'E',
    template: alertIndicatorTemplate
  };
};
