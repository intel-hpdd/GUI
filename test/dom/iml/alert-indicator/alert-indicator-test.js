import highland from "highland";
import broadcast from "../../../../source/iml/broadcaster.js";
import { default as AlertIndicator, alertIndicatorNg } from "../../../../source/iml/alert-indicator/alert-indicator.js";
import { renderToSnapshot } from "../../../test-utils.js";
import angular from "../../../angular-mock-setup.js";

describe("AlertIndicator DOM testing", () => {
  let alertStream, recordId, displayType, node, popover, i, stream, stateLabel, alerts, tooltip, template, $scope;

  beforeEach(
    angular.mock.module($compileProvider => {
      $compileProvider.component("alertIndicator", alertIndicatorNg);
    })
  );

  beforeEach(
    angular.mock.inject(($compile, $rootScope) => {
      stream = highland();

      alertStream = broadcast(stream);

      recordId = "host/6";

      displayType = "medium";

      stream.write([]);

      $scope = $rootScope.$new();
      $scope.recordId = recordId;
      $scope.displayType = displayType;
      $scope.alertStream = alertStream;
      template = '<alert-indicator record-id="recordId" display-type="displayType" alert-stream="alertStream" />';

      node = $compile(template)($scope)[0];
      $scope.$digest();

      document.body.appendChild(node);

      popover = node.querySelector.bind(node, ".popover");
      i = node.querySelector.bind(node, "i");
      alerts = node.querySelectorAll.bind(node, "li");

      stateLabel = node.querySelector.bind(node, ".state-label");

      tooltip = node.querySelector.bind(node, ".tooltip");
    })
  );

  afterEach(() => {
    $scope.$destroy();
    document.body.removeChild(node);
  });

  describe("response contains alerts", () => {
    let response;

    beforeEach(() => {
      response = [
        {
          affected: ["host/6"],
          message: "response message"
        },
        {
          affected: ["host/6"],
          message: "response message 2"
        }
      ];

      stream.write(response);
      i().click();
      $scope.$digest();
    });

    it("should have an alert message if the response contains one.", () => {
      expect(alerts()[0].textContent).toEqual("response message");
    });

    it("should add more alerts when added", () => {
      response.push({
        affected: ["host/6"],
        message: "response message 3"
      });
      stream.write(response);

      expect(alerts()[2].textContent).toEqual("response message 3");
    });

    it("should remove old alerts", () => {
      response.pop();
      stream.write(response);

      expect(alerts().length).toBe(1);
    });

    it("should hide the popover if there are no alerts", () => {
      stream.write([]);

      expect(popover()).toBeNull();
    });

    it("should show the label", () => {
      expect(stateLabel().textContent).toEqual("2 Issues");
    });
  });

  describe("exclamation icon interaction", () => {
    beforeEach(() => {
      const response = [
        {
          affected: ["host/6"],
          message: "response message"
        }
      ];

      stream.write(response);
    });

    it("should display the info icon", () => {
      expect(i()).toBeShown();
    });

    it("should display the popover after clicking info icon", () => {
      i().click();

      expect(popover()).toBeShown();
    });

    it("should display the tooltip after mousing over the info icon", () => {
      i().dispatchEvent(new MouseEvent("mouseover"));
      expect(tooltip()).toBeShown();
    });

    it("should not show the label on small size", () => {
      expect(
        renderToSnapshot(<AlertIndicator viewer={alertStream} size="small" recordId={recordId} />)
      ).toMatchSnapshot();
    });
  });
});
