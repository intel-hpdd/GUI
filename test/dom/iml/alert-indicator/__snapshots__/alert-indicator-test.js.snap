// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertIndicator DOM testing exclamation icon interaction should not show the label on small size 1`] = `
<div>
  <span
    class="record-state"
  >
    <span
      class="icon-wrap tooltip-container tooltip-hover"
      popoverbutton="true"
    >
      <i
        class="fa activate-popover fa-exclamation-circle"
      />
      
      <div
        class="tooltip inferno-tt top medium "
      >
        <div
          class="tooltip-arrow"
        />
        <div
          class="tooltip-inner"
        >
          <span>
            response message
          </span>
        </div>
      </div>
    </span>
    
    
  </span>
</div>
`;
